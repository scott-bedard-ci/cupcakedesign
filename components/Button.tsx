"use client"

import React from "react"
import { cn } from "../lib/utils"
import { createComponent } from "../lib/createComponent"
import { useTouchInteraction } from "../hooks/useTouchInteraction"
import type { IconComponentProps, SizeableComponentProps, TouchableComponentProps } from "../lib/types"

// Define button types
export type ButtonVariant = "primary" | "secondary" | "tertiary"

// Define button props
export interface ButtonProps
  extends SizeableComponentProps,
    TouchableComponentProps,
    IconComponentProps,
    React.ButtonHTMLAttributes<HTMLButtonElement> {
  /**
   * The type of button
   * @default "primary"
   */
  variant?: ButtonVariant
}

/**
 * Cupcake Button component for actions and navigation
 */
export const Button = createComponent<ButtonProps, HTMLButtonElement>(
  {
    displayName: "Button",
    defaultProps: {
      variant: "primary",
      size: "large",
      touchFeedback: true,
    },
    dataAttributes: ["variant", "size"],
  },
  (
    {
      variant = "primary",
      size = "large",
      disabled = false,
      leftIcon,
      rightIcon,
      iconOnly = false,
      className,
      children,
      touchFeedback = true,
      ...props
    },
    ref,
  ) => {
    // Determine if the button should have text
    const hasText = !iconOnly && children

    // Use touch interaction hook
    const { touchState, touchHandlers, touchClasses } = useTouchInteraction({
      feedbackEnabled: touchFeedback && !disabled,
    })

    // Size classes with improved touch targets
    const sizeClasses = {
      small: "text-sm py-1.5 px-3 rounded-md min-h-touch min-w-touch",
      medium: "text-base py-2 px-4 rounded-md min-h-touch min-w-touch",
      large: "text-base py-2.5 px-5 rounded-md min-h-touch-lg min-w-touch-lg",
    }

    // Variant classes using color tokens from Figma designs
    const variantClasses = {
      primary:
        "bg-[#1046DF] text-white border border-transparent hover:bg-[#0D3AB7] active:bg-[#0A2E92] disabled:bg-[#CCCCCC] disabled:text-[#666666] disabled:border-transparent",
      secondary:
        "bg-white border border-[#1046DF] text-[#1046DF] hover:bg-[#EBF0FF] active:bg-[#EBF0FF]/70 disabled:bg-white disabled:text-[#666666] disabled:border-[#CCCCCC]",
      tertiary:
        "bg-white border border-[#da1e28] text-[#da1e28] hover:bg-[#fff1f1] active:bg-[#fff1f1]/70 disabled:bg-white disabled:text-[#666666] disabled:border-[#CCCCCC]",
    }

    return (
      <button
        ref={ref}
        className={cn(
          "inline-flex items-center justify-center font-medium transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-[#1046DF]",
          "touch-manipulation", // Improves touch response
          sizeClasses[size],
          variantClasses[variant],
          iconOnly && "p-2.5", // Slightly larger for touch targets
          touchClasses,
          className,
        )}
        disabled={disabled}
        {...touchHandlers}
        {...props}
      >
        {leftIcon && <span className={cn("flex-shrink-0", hasText && "mr-2")}>{leftIcon}</span>}
        {hasText && <span>{children}</span>}
        {rightIcon && <span className={cn("flex-shrink-0", hasText && "ml-2")}>{rightIcon}</span>}
      </button>
    )
  },
)

// Define button group props
export interface ButtonGroupProps extends SizeableComponentProps {
  /**
   * The children (Button components)
   */
  children: React.ReactNode
  /**
   * Whether the buttons should be stacked vertically
   * @default false
   */
  vertical?: boolean
}

/**
 * Cupcake ButtonGroup component for grouping related buttons
 */
export const ButtonGroup = createComponent<ButtonGroupProps>(
  {
    displayName: "ButtonGroup",
    defaultProps: {
      vertical: false,
    },
  },
  ({ children, vertical = false, className, ...props }, ref) => {
    return (
      <div
        ref={ref}
        className={cn("flex", vertical ? "flex-col gap-3 md:gap-4" : "flex-wrap md:flex-row gap-3 md:gap-4", className)}
        {...props}
      >
        {React.Children.map(children, (child) => {
          if (React.isValidElement(child)) {
            return child
          }
          return null
        })}
      </div>
    )
  },
)
