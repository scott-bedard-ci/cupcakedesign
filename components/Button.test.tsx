"use client"
import { render, screen, fireEvent } from "@testing-library/react"
import { Button, ButtonGroup } from "./Button"
import { testComponentVariants } from "../test/testUtils"

describe("Button Component", () => {
  // Rendering tests
  test("renders with default props", () => {
    render(<Button>Click me</Button>)
    const button = screen.getByRole("button", { name: "Click me" })
    expect(button).toBeInTheDocument()
    expect(button).toHaveAttribute("data-cupcake-variant", "primary")
    expect(button).toHaveAttribute("data-cupcake-size", "large")
  })

  // Variant tests
  test("renders variants correctly", () => {
    testComponentVariants(Button, "variant", ["primary", "secondary", "tertiary"], {
      children: "Test Button",
    })
  })

  // Size tests
  test.each(["small", "medium", "large"] as const)("renders %s size correctly", (size) => {
    render(<Button size={size}>{size} button</Button>)
    const button = screen.getByRole("button", { name: `${size} button` })
    expect(button).toHaveAttribute("data-cupcake-size", size)
  })

  // Disabled state test
  test("renders disabled state correctly", () => {
    render(<Button disabled>Disabled button</Button>)
    const button = screen.getByRole("button", { name: "Disabled button" })
    expect(button).toBeDisabled()
  })

  // Icon tests
  test("renders with left icon", () => {
    render(<Button leftIcon={<span data-testid="left-icon" />}>Button with left icon</Button>)

    expect(screen.getByTestId("left-icon")).toBeInTheDocument()
    expect(screen.getByText("Button with left icon")).toBeInTheDocument()
  })

  test("renders with right icon", () => {
    render(<Button rightIcon={<span data-testid="right-icon" />}>Button with right icon</Button>)

    expect(screen.getByTestId("right-icon")).toBeInTheDocument()
    expect(screen.getByText("Button with right icon")).toBeInTheDocument()
  })

  test("renders icon-only button", () => {
    render(<Button leftIcon={<span data-testid="icon" />} iconOnly aria-label="Icon button" />)

    expect(screen.getByTestId("icon")).toBeInTheDocument()
    expect(screen.getByRole("button", { name: "Icon button" })).toBeInTheDocument()
  })

  // Event tests
  test("calls onClick when clicked", () => {
    const handleClick = jest.fn()
    render(<Button onClick={handleClick}>Clickable button</Button>)

    const button = screen.getByRole("button", { name: "Clickable button" })
    fireEvent.click(button)

    expect(handleClick).toHaveBeenCalledTimes(1)
  })

  // Accessibility tests
  test("has appropriate ARIA attributes for icon-only button", () => {
    render(<Button leftIcon={<span data-testid="icon" />} iconOnly aria-label="Accessible icon button" />)

    const button = screen.getByRole("button", { name: "Accessible icon button" })
    expect(button).toBeInTheDocument()
  })

  // Snapshot test
  test("matches snapshot", () => {
    const { container } = render(
      <Button variant="primary" size="large">
        Snapshot Test
      </Button>,
    )

    expect(container).toMatchSnapshot()
  })
})

describe("ButtonGroup Component", () => {
  // Rendering tests
  test("renders button group with children", () => {
    render(
      <ButtonGroup>
        <Button data-testid="button-1">Button 1</Button>
        <Button data-testid="button-2">Button 2</Button>
      </ButtonGroup>,
    )

    expect(screen.getByTestId("button-1")).toBeInTheDocument()
    expect(screen.getByTestId("button-2")).toBeInTheDocument()
  })

  test("renders vertical button group", () => {
    render(
      <ButtonGroup vertical data-testid="button-group">
        <Button>Button 1</Button>
        <Button>Button 2</Button>
      </ButtonGroup>,
    )

    const buttonGroup = screen.getByTestId("button-group")
    expect(buttonGroup).toHaveClass("flex-col")
  })

  // Snapshot test
  test("matches snapshot", () => {
    const { container } = render(
      <ButtonGroup>
        <Button variant="primary">Primary</Button>
        <Button variant="secondary">Secondary</Button>
      </ButtonGroup>,
    )

    expect(container).toMatchSnapshot()
  })
})
