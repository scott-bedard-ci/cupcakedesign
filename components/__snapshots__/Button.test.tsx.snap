// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component matches snapshot 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center font-medium transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-[#1046DF] touch-manipulation text-base py-2.5 px-5 rounded-md min-h-touch-lg min-w-touch-lg bg-[#1046DF] text-white border border-transparent hover:bg-[#0D3AB7] active:bg-[#0A2E92] disabled:bg-[#CCCCCC] disabled:text-[#666666] disabled:border-transparent"
    data-cupcake-component="button"
    data-cupcake-size="large"
    data-cupcake-variant="primary"
  >
    <span>
      Snapshot Test
    </span>
  </button>
</div>
`;

exports[`ButtonGroup Component matches snapshot 1`] = `
<div>
  <div
    class="flex flex-wrap md:flex-row gap-3 md:gap-4"
    data-cupcake-component="buttongroup"
  >
    <button
      class="inline-flex items-center justify-center font-medium transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-[#1046DF] touch-manipulation text-base py-2.5 px-5 rounded-md min-h-touch-lg min-w-touch-lg bg-[#1046DF] text-white border border-transparent hover:bg-[#0D3AB7] active:bg-[#0A2E92] disabled:bg-[#CCCCCC] disabled:text-[#666666] disabled:border-transparent"
      data-cupcake-component="button"
      data-cupcake-size="large"
      data-cupcake-variant="primary"
    >
      <span>
        Primary
      </span>
    </button>
    <button
      class="inline-flex items-center justify-center font-medium transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-[#1046DF] touch-manipulation text-base py-2.5 px-5 rounded-md min-h-touch-lg min-w-touch-lg bg-white border border-[#1046DF] text-[#1046DF] hover:bg-[#EBF0FF] active:bg-[#EBF0FF]/70 disabled:bg-white disabled:text-[#666666] disabled:border-[#CCCCCC]"
      data-cupcake-component="button"
      data-cupcake-size="large"
      data-cupcake-variant="secondary"
    >
      <span>
        Secondary
      </span>
    </button>
  </div>
</div>
`;
